execution_engine/src/core/runtime/mod.rs

```diff
@@ -3345,6 +3346,84 @@ where
         self.gas(cost)?;
         Ok(())
     }
+
+    fn new_dictionary(&mut self, output_size_ptr: u32) -> Result<Result<(), ApiError>, Error> {
+        // Proceed with creating new URefs
+        let new_uref = self.context.new_unit_uref()?;
+
+        // check we can write to the host buffer
+        if let Err(err) = self.check_host_buffer() {
+            return Ok(Err(err));
+        }
+        // create CLValue for return value
+        let new_uref_value = CLValue::from_t(new_uref)?;
+        let value_size = new_uref_value.inner_bytes().len();
+        // write return value to buffer
+        if let Err(err) = self.write_host_buffer(new_uref_value) {
+            return Ok(Err(err));
+        }
+        // Write return value size to output location
+        let output_size_bytes = value_size.to_le_bytes(); // Wasm is little-endian
+        if let Err(error) = self.memory.set(output_size_ptr, &output_size_bytes) {
+            return Err(Error::Interpreter(error.into()));
+        }
+
+        Ok(Ok(()))
+    }
+
+    /// Similar to `read`, this function is for reading from the "local cluster" of global state
+    /// referenced by the uref.
+    fn dictionary_get(
+        &mut self,
+        uref_ptr: u32,
+        uref_size: u32,
+        key_bytes_ptr: u32,
+        key_bytes_size: u32,
+        output_size_ptr: u32,
+    ) -> Result<Result<(), ApiError>, Trap> {
+        if !self.can_write_to_host_buffer() {
+            // Exit early if the host buffer is already occupied
+            return Ok(Err(ApiError::HostBufferFull));
+        }
+
+        let uref: URef = self.t_from_mem(uref_ptr, uref_size)?;
+        let key_bytes = self.bytes_from_mem(key_bytes_ptr, key_bytes_size as usize)?;
+
+        let cl_value = match self.context.dictionary_get(uref, &key_bytes)? {
+            Some(cl_value) => cl_value,
+            None => return Ok(Err(ApiError::ValueNotFound)),
+        };
+
+        let value_size = cl_value.inner_bytes().len() as u32;
+        if let Err(error) = self.write_host_buffer(cl_value) {
+            return Ok(Err(error));
+        }
+
+        let value_bytes = value_size.to_le_bytes(); // Wasm is little-endian
+        if let Err(error) = self.memory.set(output_size_ptr, &value_bytes) {
+            return Err(Error::Interpreter(error.into()).into());
+        }
+
+        Ok(Ok(()))
+    }
+
+    /// Writes `value` under a key derived from `key` in the dictionary partition of GlobalState
```

Comment:
If I didn't already know what you meant, this comment would be useless to me. Consider expounding on it a bit to be less inscrutable?

Reply:
Addressed